### ISSUE TITLE
Options to change parent link behaviour in responsive dropdown menus

### ISSUE THREAD

**Issue by BWPanda** [backdrop/backdrop-issues#2370]
This is a follow-up of https://github.com/backdrop/backdrop-issues/issues/2107, extracted from https://github.com/backdrop/backdrop-issues/issues/2367.

In a dropdown menu that responds to smaller screens/devices, when you have a parent link that you want to click on you have to essentially double-click on it - once to expand the child menu, and then again to navigate to the parent link's URL. At the same time (when clicking the parent link), the child menu disappears, so it appears to be that clicking the parent link the second time simply closes the child menu, that is until the page reloads and you're taken to the parent's URL.

I understand that this is what some people are used to having in a responsive menu, so I'd like to suggest at least having the option to change this behaviour to one of the following:

1. Clone the parent link into the child menu so that the parent link simply becomes an open/close button for the child menu and the parent's URL is accessed through the cloned link.
2. Make the parent link link only to it's URL, and leave the opening/closing of the child menu to the +/- sign on the right.

---

ADVOCATE: @laryn 

--

**Comment by olafgrabienski** [comment]
Yes, depending on the loading time of a page and other factors, the current situation can be confusing for users. In my opinion one of the mentioned options, or both of them, would improve the situation.

On the one hand I prefer the first option (clone the parent link) because a user doesn't have to think about where to click: on the link vs. on the +/- signs.

On the other hand the second option (+/- for opening and closing) looks more elegant.

What do others think about it?

**Comment by klonos** [comment]
I think that both options have very valid use cases and this is subject to preference as well. I say provide both of them as two options and set one as the default (we can bikeshed which one, but even randomly picking one would suffice if we provide people the option to change). When we have metrics in place, we can reevaluate which one should be the default.

With the +/- signs, I think that we should make them more prominent (different shade of background color perhaps?) if they are set to handle the expanding/collapsing of the menu. Otherwise, we risk people hitting a UX WTF until they notice them (since they are on the side of the screen and are less prominent than the menu item text).

**Comment by olafgrabienski** [comment]
> With the +/- signs, I think that we should make them more prominent if they are set to handle the expanding/collapsing of the menu (different shade of background color perhaps?).

Good point! See for instance the [demo page of Mean menu](http://meanthemes.wpengine.com/demo/meanmenu/demo.html) where the part with the +/- links is separated by borders.

**Comment by klonos** [comment]
Yep, that's exactly what I meant. For future reference (because things on the net are not always for ever), here's a screenshot from the demo page @olafgrabienski linked to in his comment above:

![responsive_menu-different_background_color_on_expand-collapse_handles](https://cloud.githubusercontent.com/assets/2423362/21087296/c07db338-c078-11e6-8520-7059eec58ed5.png)

**Comment by olafgrabienski** [comment]
@jenlampton I saw your [Smartmenus module on drupal.org](https://www.drupal.org/project/smartmenus) and had a look at the initial sponsor website, [Sirius Puppy Training](https://www.siriuspup.com). Is the opening and closing behaviour of the +/- signs on that site (which we discussed above for Backdrop) part of the Drupal module?

**Comment by jenlampton** [comment]
I'm not sure exactly what you are asking, but I haven't added anything other than CSS to that site, everything else is in the module.

**Comment by olafgrabienski** [comment]
Thanks @jenlampton! Actually, I had the impression the menu on that site behaved different than the Backdrop responsive dropdown, but it doesn't.

**Comment by olafgrabienski** [comment]
> With the +/- signs, I think that we should make them more prominent if they are set to handle the expanding/collapsing of the menu - @klonos in https://github.com/backdrop/backdrop-issues/issues/2370#issuecomment-266259217

Interestingly, the +/- signs **are** already set to handle the expanding/collapsing of a child menu.

The parent link itself expands and collapses the child menu as well. Additionally, it works as a link to the parent page, but only when it's expanded, which is quite confusing.

I suggest the following changes to fix this issue:

- Make the parent link link only to it's URL, and leave the opening/closing of the child menu to the +/- sign on the right (cf. suggestion 2. in the [issue description](https://github.com/backdrop/backdrop-issues/issues/2370#issue-192421960)).
- Style the +/- signs more prominent (as an example see screenshot in @klonos' [comment above](https://github.com/backdrop/backdrop-issues/issues/2370#issuecomment-266336547)).
- Additionally, indent the child menus via CSS. (Should be a separate issue, I guess.)

The styling of the +/- signs should be easy, will try to do that.

Regarding the parent link behaviour: don't know how to change it. Anybody else?

**Comment by Dinornis** [comment]
There is a pull request at #2553 that is addressing this issue.

**Comment by BWPanda** [comment]
So the PR in https://github.com/backdrop/backdrop-issues/issues/2553 made it into core, but it only affects the Basis theme. Perhaps we can do something similar here that's theme-agnostic...?

**Comment by jenlampton** [comment]
If we do, I don't think it should include the indent for child-items. That's not going to work for menus that have many levels of depth, as the link width will get too short to hold all of the link text.

**Comment by BWPanda** [comment]
I tried making a PR for this, but it wasn't working for me... Based on what I found here (https://www.smartmenus.org/docs/#collapsibleBehavior) it seems that we'd just need to set `collapsibleBehavior` to `link`, but I couldn't work out where to do this.

**Comment by laryn** [comment]
Our accessibility expert called out the need for two taps on mobile as an a11y issue. I've just added a PR that exposes the `collapsibleBehavior` setting as an option. @BWPanda are you interested in testing?

**Comment by stpaultim** [comment]
I added this to the milestone, because @laryn offered to advocate for it in Zulip today.

**Comment by laryn** [comment]
Also, if anybody sees a good way to change the default option without affecting existing sites, I'm all ears.

**Comment by laryn** [comment]
I started out putting the descriptions in there but it looked too wordy. Maybe we could add the descriptions from the library (add here: https://github.com/backdrop/backdrop-issues/issues/2370#issuecomment-1507533757 ) onto a docs page and link to that?

**Comment by olafgrabienski** [comment]
Are the `accordeon` options that important? I find them not very understandable at first sight, and not adding them would also save a lot of space for descriptions.

**Comment by laryn** [comment]
I think I see one other change that should be made: it currently shows if the "Display menu toggle button on small screens" is checked, but should also probably only show for "Dropdown" menu style (not for hierarchical tree or top level only).

**Comment by laryn** [comment]
I updated the PR:

<img width="856" alt="image" src="https://user-images.githubusercontent.com/871421/234333126-e644a3ac-096a-4be4-9da3-3f2b27e8ea0c.png">

**Comment by yorkshire-pudding** [comment]
I think the accordion options can have benefits on more complex menus. If previous trees are left open, then they can be confusing.  While good UI should try to avoid complex menus, we can't dictate what site architects might build so allowing these options gives flexibility.

**Comment by olafgrabienski** [comment]
@laryn Thanks for the updated PR! It works for me, but I had to clear caches so that a new behavior was applied for anonymous visitors.

@yorkshire-pudding Do you have an idea regarding the user interface if we provide all the options? In my opinion, they are quite overwhelming and not easy to understand at first sight.

**Comment by laryn** [comment]
Interesting idea. I'd maybe put the "Collapsible Behavior" option above, and then a checkbox below which is unchecked by default and  says something like:

  [  ] Accordion-style: also resets any visible sub menus from deeper levels or other branches.

**Comment by laryn** [comment]
Thank goodness for "squash and merge" 😅  but next draft is ready:

<img width="715" alt="image" src="https://user-images.githubusercontent.com/871421/234626565-e2a1f23e-c848-4e69-a43f-62d55a73c738.png">

**Comment by olafgrabienski** [comment]
> next draft is ready

Thanks again for the update! Tested in the PR sandbox, it works for me, and I like the interface.

As before, I didn't see a change immediately as anonymous visitor, only after an admin cleared caches.

**Comment by laryn** [comment]
@herbdool Are you able to do another code review since we've modified some things?

**Comment by herbdool** [comment]
It's getting close.

I find the wording a bit confusing: "Also resets any visible sub menus from deeper levels or other branches upon expanding."

Maybe just make it: "Close all sub-menus when closing the parent menu items." If that is close enough to correct.

I'm wondering too if it'll look nicer if the three options were radio buttons so that the descriptions can be proper descriptions per radio button.

**Comment by laryn** [comment]
@herbdool @olafgrabienski Here's the latest:

<img width="765" alt="image" src="https://user-images.githubusercontent.com/871421/235284321-a122abf7-2883-4100-a7a8-bde3caa918fb.png">

**Comment by herbdool** [comment]
I think that improves it. I'll RTBC.

I have a dumb question: so these options are only for mobile? They don't change the way the full menu works? I guess not, because the opens just on mouseover not clicks.

**Comment by laryn** [comment]
@herbdool Yes, these are just for the behaviour when the menu is collapsed/mobile. There is a setting that affects the full menu to turn it to a click drop instead of a hover drop, and a PR that actually cleans up that implementation is ready for review, too. In case you have menus on the brain and want to review another. ;) 

https://github.com/backdrop/backdrop-issues/issues/4488

**Comment by olafgrabienski** [comment]
Nice, a great improvement!

@laryn I also have the impression that clearing caches isn't needed anymore.

**Comment by klonos** [comment]
@laryn sorry for the late review here - I've only tested the UI/UX aspect of this, and I have the following feedback and suggestions (if others agree):
- Make the "Parent item's behavior in collapsible (mobile) view." description the title of the radios, instead of the current "Collapsible Behavior" (which shouldn't have a capitalized "B" in any case BTW). Perhaps also slightly change it to "Parent item behavior in collapsible/mobile mode"(?).
- Add the new `#indentation` property to the radios, so that they are "nested" under the "Display menu toggle button on small screens" checkbox on which they depend.
- Change the "Parent functions as toggle without a link." description to something that mentions what is being toggled. Perhaps "Parent functions only as an expand/collapse toggle for its children. No link functionality."
- Similarly, change the "Parent text is a link, +/- acts as a toggle." to something like "Parent only functions as a menu link. Separate +/- button acts as an expand/collapse toggle." instead.

**Comment by klonos** [comment]
> - Add the new `#indentation` property to the radios, ...

That's broken ^^ ...I'll file a separate issue for it shortly.

**Comment by klonos** [comment]
...follow-up here: #6090

**Comment by klonos** [comment]
This is a side-by-side of the current PR sandbox vs. the best I could come up with on my local:
![image](https://user-images.githubusercontent.com/2423362/235353091-679fa83c-dde4-4c7c-b43b-792b8b5e2b64.png)

@laryn if you and others find it to be an improvement, you can create a PR against yours branch from mine: https://github.com/laryn/backdrop/compare/1.x-issue-2370...klonos:backdrop:issue-2370 ...you will need to rebase your branch though - it seems to be behind current 1.x. My changes are limited to only the `core/modules/system/system.menu.inc` file.

**Comment by klonos** [comment]
@backdrop/core-committers please feel free to merge the current PR as is, so that this really nice feature can get into the next minor release. My comments/suggestions can be reviewed and this can be tweaked further in the two weeks between the pre-release and the final release 😉

**Comment by quicksketch** [comment]
We discussed this in this past week's dev meeting. At the time, I wasn't sure about it. But in using it I immediately saw the value in the behavior choice. For menus, it definitely would be nice if this was a site-wide setting instead of being per-page, but that's true of all the menu settings. For what it's worth, I like the existing short labels.

**Comment by quicksketch** [comment]
I merged https://github.com/backdrop/backdrop/pull/4402 into 1.x for 1.25.0. Thanks everyone!

[01595f2](https://github.com/backdrop/backdrop/commit/01595f29524899188a42d4de15b31499291697f1) by @laryn, @klonos, @BWPanda, @herbdool, @olafgrabienski, and @yorkshire-pudding.

**Comment by klonos** [comment]
Follow-up here: #6093
